### DISCUSSION TITLE
Large raster file with locate_trees()

### DISCUSSION THREAD

**Discussion by bengwill** [r-lidar/lidR#682] [Q&A]
Hello Jean and the rest of the community. I am working with a large raster file, a canopy height model of a small town, to use the locate_trees algorithm on.

It continues to report the error "Error: Large on-disk rasters are not supported by locate_tree. Load the raster manually."

I have tried tiling my raster with the makeTiles() function and then reading it in as a .vrt but that returns the same error.

Could I please get some guidance on how I can process this raster with the locate_tree function and not running out of memory?

Thank you very much for your time and assistance.

**Comment by bengwill** [comment]
So far, I have just separated my CHM into four separate smaller rasters and am processing them individually. 

It seems to be working, but I am wondering if there is a better way to do an analysis on the whole data set at once so that the incremental output would be continuous?

**Reply by Jean-Romain** [reply]
lidR is a point cloud oriented software. Here you are asking for a raster oriented tools. lidR provides little to process large raster you have to do it by your own. If you think your chunks are small enough to be loaded in memory you can multiply by 1 to get it loaded in memory. `terra` does not load in memory by default. lidR loads on the fly if the raster is small but apparently in your case it is not small enough

**Reply by bengwill** [reply]
Jean-Romain, thank you very much for your prompt reply! 

Indeed, these tools are meant for LiDAR and I am currently clipping out my LAS dataset to have it processed through locate_trees. We wanted to at least compare with the CHM route and see how well they each work.

Chunking up my rasters manually did the trick and I was able to process each individually. Maybe one day someone will release catalog functions for rasters like you have in LidR for lidar :)

Thank you for all of your hard work and dedication to this project, its something we are all very proud to be using.

**Reply by Jean-Romain** [reply]
Well actually the `LAScatalog` engine can process chunks of raster with buffer if you trick a bit the engine. This would looks like that

```r
ctg = readLAScatalog()
rast = rast() # big raster

myfun = function(chunk, rast)
{
  bbox = extent(chunk)
  bbox = bbox + buffer
  rast_in_memory = crop(rast, bbox)
  # process rast_in_memory
  rast_in_memory = crop(rast_in_memory, bbox)
  return(rast_in_memory)
}

opt_output_file(ctg) = "..."
opt_chunk_size(ctg) = 1000
opt_wall_to_wall(ctg) = FALSE # This one desactivates some control especially you are not reading any point cloud in your function.
catalog_sapply(ctg, myfun, rast = rast)
```

**Reply by DeltaForest-DP** [reply]
Hi Jean-Romain, I am having the same problem - I want to locate_trees on large CHM rasters... I'd like to process as if they were in a catalog, so I like your code above - EXCEPT on the very first line (readLAScatalog()) there are no las/laz files to reference (only CHMs), so I get an error here.  How do you trick the catalog engine here??

Thanks, DP

**Reply by Jean-Romain** [reply]
The lidR package is designed to process point clouds and has an engine to process large point clouds. lidR is not designed to process large rasters.
